cluster.go

```diff
@@ -422,17 +422,22 @@ func (cluster *mongoCluster) syncServersLoop() {
 	debugf("SYNC Cluster %p is stopping its sync loop.", cluster)
 }
 
-func (cluster *mongoCluster) server(addr string, tcpaddr *net.TCPAddr) *mongoServer {
+func (cluster *mongoCluster) server(addr string, raddr net.Addr) *mongoServer {
 	cluster.RLock()
-	server := cluster.servers.Search(tcpaddr.String())
+	server := cluster.servers.Search(raddr.String())
 	cluster.RUnlock()
 	if server != nil {
 		return server
 	}
-	return newServer(addr, tcpaddr, cluster.sync, cluster.dial, cluster.dialInfo)
+	return newServer(addr, raddr, cluster.sync, cluster.dial, cluster.dialInfo)
 }
 
-func resolveAddr(addr string) (*net.TCPAddr, error) {
+func resolveAddr(addr string) (net.Addr, error) {
+	// Check if addr is a Unix Domain Socket
+	if strings.HasSuffix(addr, ".sock") {
```

Comment:
This may not work, people can and will rename their AF_UNIX sockets at times. You could os.Stat(addr).

Reply:
(or simply search for a "/" in the string)

Reply:
https://github.com/pzillmann/mgo/blob/801285371e4ffb7ed938b450a657ac2e9d68ce72/session.go#L235

```
UNIX domain sockets MUST end in ".sock" and MUST be URL encoded, for
example:

	mongodb://user:pass@%2Ftmp%2Fmongodb-27017.sock/authDB?replicaSet=rs

The host information cannot contain an unescaped slash ("/"), if it does then an
exception MUST be thrown informing users that paths must be URL encoded. For
example:

	Unsupported host '/tmp/mongodb-27017.sock', UNIX socket domain paths must be
	URL encoded.
```

A socket in your local directory could be in some strange way in $PATH that you can use a socket without any directory-seperator.

Reply:
Ah, I see, I wasn't aware of this mgo imposed naming convention.